mhkit/wave/io.py

```diff
@@ -105,3 +109,303 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
+    assert isinstance(proxy , (str, type(None))), 'If specified proxy must be a string'
+    supported =__supported_ndbc_params(parameter)
+    if buoy_number != None:
+        assert len(buoy_number) == 5, ' 5-character alpha-numeric station identifier'
+    
+    ndbc_data = f'https://www.ndbc.noaa.gov/data/historical/{parameter}/'
+    if proxy == None:
+        response = requests.get(ndbc_data)
+    else:
+        response = requests.get(ndbc_data, proxies=proxy)
+    
+    status = response.status_code 
+    if status != 200:
+        msg=f"request.get{ndbc_data} failed by returning code of {status}"
+        raise Exception(msg)            
+
+
+    filenames = pd.read_html(response.text)[0].Name.dropna()    
+    buoys = _ndbc_parse_filenames(parameter, filenames)
+
+    available_data = buoys.copy(deep=True)
+    if buoy_number != None:
+        available_data = buoys[buoys.id==buoy_number]
+        
+    return available_data
+    
+def _ndbc_parse_filenames(parameter, filenames):  
+    '''
+    Takes a list of available filenames as a series from NDBC then 
+    parses out the station ID and year from the file name.
+    
+    Parameters
+    ----------
+    parameter: string
+        'swden'	:	'Raw Spectral Wave Current Year Historical Data'
+        'stdmet':   'Standard Meteorological Current Year Historical Data'
+    filenames: Series
+        List of compressed file names from NDBC
+     
+    Returns
+    -------
+    buoys: DataFrame
+        DataFrame with keys=['id','year','file_name']    
+    '''  
+    assert isinstance(filenames, pd.Series), 'filenames must be of type pd.Series' 
+    assert isinstance(parameter, str), 'parameter must be a string'
+    supported =__supported_ndbc_params(parameter)
+    
+    file_seps = {
+                'swden' : 'w',
+                'stdmet' : 'h'
+               }
+    file_sep= file_seps[parameter]
+    
+    filenames = filenames[filenames.str.contains('.txt.gz')]
+    buoy_id_year_str = filenames.str.split('.', expand=True)[0]
+    buoy_id_year = buoy_id_year_str.str.split(file_sep, n=1,expand=True)
+    buoys = buoy_id_year.rename(columns={0:'id', 1:'year'})
+    
+    expected_station_id_length = 5
+    buoys = buoys[buoys.id.str.len() == expected_station_id_length]
+    buoys['filename'] = filenames  
+    return buoys    
+    
+    
+def fetch_ndbc(parameter, filenames, proxy=None):
+    '''
+    Requests data and returns a DataFrame for each filename passed
+        
+    Parameters
+    ----------
+    parameter: string
+        'swden'	:	'Raw Spectral Wave Current Year Historical Data'
+        'stdmet':   'Standard Meteorological Current Year Historical Data'
+    filenames: DataFrame
+	    Data filenames on https://www.ndbc.noaa.gov/data/historical/{parameter}/
+	proxy: string
+	    Proxy URL   
+        
+    Returns
+    -------
+    ndbc_data: dict
+        Dictionary of NDBC data 
+    '''
```

Comment:
Okay, so I played with this a little bit, and it's pretty simple to convert one or two years of data to a DataFrame, but what about all  historical data? I think it would be useful to make a DataFrame containing all historical data the default.

``` python
df1 = pd.DataFrame.from_dict(ndbc_data['2019'])
df2 = pd.DataFrame.from_dict(ndbc_data['2018'])
new =pd.concat([df1,df2])
```

then you should be able to run functions from `mhkit.wave.resource`
```python
mhkit.wave.resource.significant_wave_height(new['2019-01-01 00:00:00'],frequency_bins=new.columns.values)
```
but I'm getting a datetime error
"Invalid comparison between dtype=datetime64[ns] and int"

Reply:
```
Create DataFrame for all May data, plot May data
may_data = new[new.index.month==5]
may_data.mean().plot()
may_data[:5].T.plot()

Plot all data from first day in May
may_data[may_data.index.day==1].T.plot()
```

Reply:
> Why not return `ndbc_data` as dictionary instead of a DataFrame? I think this return should be the same as `read_NDBC_file`?
> 
> I re-read the description and now I see that
> "`fetch_ndbc`: Takes filenames and returns a dictionary of DataFrames indexed by years for each filename passed"
> 
> What is the reasoning behind this? Why not just output a DataFrame that includes all data?
> 
> ```
> ndbc_data['2019'].head()
> ndbc_data['2001'].plot()
> ```
> 
> It's simple to combine years of data to a DataFrame, but what about all historical data? I think it would be useful to make a DataFrame containing all historical data the default.
> 
> ```python
> new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
> ```
> 
> then you should be able to run functions from `mhkit.wave.resource`
> 
> ```python
> mhkit.wave.resource.significant_wave_height(new['2019-01-01 00:00:00'],frequency_bins=new.columns.values)
> ```
> 
> but I'm getting a datetime error
> "Invalid comparison between dtype=datetime64[ns] and int"

Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g  `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).  

In your next step, I think having the inconsistent frequency bins gets even messier as you specify the frequency_bins to be the column values. To use the Hs function you would currently need to transpose the data. I cannot reproduce your error message.

Reply:
> Create DataFrame for all May data, plot May data
> 
> ```python
> may_data = new[new.index.month==5]
> may_data.mean().plot()
> may_data[:5].T.plot()
> ```
> 
> Plot all data from first day in May
> 
> ```python
> may_data[may_data.index.day==1].T.plot()
> ```
> 
> Resample data on annual, monthly or weekly basis
> 
> ```python
> new.resample('AS').mean().head()
> new.resample('MS').mean().head()
> new.resample('W').mean().head()
> ```

Yes having the index as a Datetime for historical data allows for this sort of analysis. I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate

Reply:
> Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).

I figured that was the motivation. How often do the frequency bins and ranges change (e.g. does it happen once due to a NDBC change, or is it somewhat random)? Would it be useful to combine all data with the same frequency bins and range into one DataFrame (even if it's not the entire data history)? Or, is it better to keep the output like this for now, and develop a separate way to merge historical data with different frequency bins and ranges?

Reply:
>  I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate

Yes, my thought was to include using pandas to view/analyze historical data as an example. This could be a new section or appendix do an existing example, or a new example. And I'm open to revisiting the conversation about indexing and/or adding a new function.

Reply:
> > Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).
> 
> I figured that was the motivation. How often do the frequency bins and ranges change (e.g. does it happen once due to a NDBC change, or is it somewhat random)? Would it be useful to combine all data with the same frequency bins and range into one DataFrame (even if it's not the entire data history)? Or, is it better to keep the output like this for now, and develop a separate way to merge historical data with different frequency bins and ranges?

So as far as I have seen the bins change once (e.g. 2 distinct frequency binning). However, the date format changes twice (3 distinct date formats).  So if you wanted to combine across years on the subset of evenly spaced bins I see this as a separate function. The user would first need to have the data and convert the index to datetime and drop the original ndbc date columns (steps 2 & 3 as covered in the example). I have already given a number of reasons why I think these should be separate steps (summary calculations are separate from requests, should we manipulate the underlying dataframe, what if the user has custom functions which expect the data in the ndbc reported format?).  Then the user could combine the subset of evenly spaced frequency bins but then the naming of the dataframe is weird for the dictionary of dataframes E.g. key value would now be '1996-2004' and '2004b-2019'. I don't see writing this to be a built-in function as explicitly useful. Where as I do think the next step would be to determine the best method to combine uneven binning and ranges and add a function for that so that we would have one object.

Reply:
> > I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate
> 
> Yes, my thought was to include using pandas to view/analyze historical data as an example. This could be a new section or appendix do an existing example, or a new example. And I'm open to revisiting the conversation about indexing and/or adding a new function.

Okay I will keep this function to return Datetime and once we reach an agreement bring it up with Kate

Reply:
> In your next step, I think having the inconsistent frequency bins gets even messier as you specify the frequency_bins to be the column values. To use the Hs function you would currently need to transpose the data. I cannot reproduce your error message.

Thanks for the reminder about indexing by frequency, not DateTime. When I transpose the data I get a different error
``` python
new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
mhkit.wave.resource.significant_wave_height(new.T)
```
_TypeError: '>' not supported between instances of 'str' and 'int'_


Then I tried specifying the frequency spacing, I get the same error
``` python
new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
 mhkit.wave.resource.significant_wave_height(new.T,frequency_bins=new.columns.values)
```
_TypeError: '>' not supported between instances of 'str' and 'int'_

Reply:
So the columns are strings of frequency bins. you would need to transpose and covert the index I believe.